### PR TITLE
`allOf` support

### PR DESCRIPTION
**Pull request by n8chur**
Additionally moves test fixture to a separate file, rather than being defined in-line.

### CODE REVIEW

File: Sources/AllOfSchema.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import ObjectMapper
+
+public struct AllOfSchema {
+    public let metadata: Metadata
+
```
Comment (AttilaTheFun) [inline Sources/AllOfSchema.swift]: I usually don't add extra newlines unless there is a comment.
E.g.:

```
public struct AllOfSchema {
    public let metadata: Metadata
    public let schemas: [Schema]
}
```

or

```
/// Something about AllOfSchema
public struct AllOfSchema {
    /// Something about metadata
    public let metadata: Metadata

    /// Something about schemas
    public let schemas: [Schema]
}
```

File: Sources/AllOfSchema.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    
+    typealias Building = AllOfSchema
+    
+    let metadata: MetadataBuilder
+    let schemaBuilders: [SchemaBuilder]
+    
+    init(map: Map) throws {
+        metadata = try MetadataBuilder(map: map)
+        
+        let allOf: [[String:Any]] = try map.value("allOf")
```
Comment (AttilaTheFun) [inline Sources/AllOfSchema.swift]: I don't believe this is necessary. [Mappable] is also Mappable.
I believe you can just do:
```
schemaBuilders = try map.value("allOf")
```

File: Sources/AllOfSchema.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            .flatMap { map -> SchemaBuilder? in
+                guard let schemaBuilder = try? SchemaBuilder(map: map) else {
+                    return nil
+                }
+                return schemaBuilder
+            }
+    }
+    
+    func build(_ swagger: SwaggerBuilder) throws -> AllOfSchema {
+        let schemas = try schemaBuilders.map {try $0.build(swagger)}
```
Comment (AttilaTheFun) [inline Sources/AllOfSchema.swift]: Spacing around curly braces:
```
{ try $0.build(swagger) }
```

File: Tests/SwaggerParserTests/Fixtures/test_all_of.json [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      }
+    }
+  },
+  "definitions": {
+    "TestAllOfBase": {
+      "required": [
+        "base",
+        "test_type"
+      ],
+      "discriminator": "test_type",
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/Fixtures/test_all_of.json]: You've added a discriminator here but not support for it. Do you want to add that in a separate PR?
If it's not going to be added in this PR I would remove the line from the fixture.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 5]
```diff
@@ -1,20 +1,129 @@
...
...lready factors in the surge multiplier.\"}}},\"Profile\":{\"properties\":{\"first_name\":{\"type\":\"string\",\"description\":\"First name of the Uber user.\"},\"last_name\":{\"type\":\"string\",\"description\":\"Last name of the Uber user.\"},\"email\":{\"type\":\"string\",\"description\":\"Email address of the Uber user\"},\"picture\":{\"type\":\"string\",\"description\":\"Image URL of the Uber user.\"},\"promo_code\":{\"type\":\"string\",\"description\":\"Promo code of the Uber user.\"}}},\"Activity\":{\"properties\":{\"uuid\":{\"type\":\"string\",\"description\":\"Unique identifier for the activity\"}}},\"Activities\":{\"properties\":{\"offset\":{\"type\":\"integer\",\"format\":\"int32\",\"description\":\"Position in pagination.\"},\"limit\":{\"type\":\"integer\",\"format\":\"int32\",\"description\":\"Number of items to retrieve (100 max).\"},\"count\":{\"type\":\"integer\",\"format\":\"int32\",\"description\":\"Total number of items available.\"},\"history\":{\"type\":\"array\",\"items\":{\"$ref\":\"#/definitions/Activity\"}}}},\"Error\":{\"properties\":{\"code\":{\"type\":\"integer\",\"format\":\"int32\"},\"message\":{\"type\":\"string\"},\"fields\":{\"type\":\"string\"}}}}}"
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift:4]: Thanks for setting up the fixtures! They're much easier to use :)

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 6]
```diff
@@ -1,20 +1,129 @@
...
-            print(error)
+    var testFixtureFolder: URL!
+    
+    override func setUp() {
+        testFixtureFolder = URL(fileURLWithPath: #file).deletingLastPathComponent().appendingPathComponent("Fixtures")
+    }
+    
+    func testInitialization() {
+        let url = testFixtureFolder.appendingPathComponent("uber.json")
+        let jsonString = try! NSString(contentsOfFile: url.path, encoding: String.Encoding.utf8.rawValue) as String
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: You can just use 
```String.init(contentsOf: URL, encoding: String.Encoding)```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 7]
```diff
@@ -1,20 +1,129 @@
...
+        let swagger = try! Swagger(JSONString: jsonString)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: I would prefer we didn't crash the test if any one of these try! calls failed.
Can you wrap the whole test in a try / catch block and just print the error + fail the test in that case?

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 8]
```diff
@@ -1,20 +1,129 @@
...
+        
+        XCTAssertEqual(swagger.host?.absoluteString, "api.uber.com")
+    }
+    
+    func testAllOfSupport() {
+        let url = testFixtureFolder.appendingPathComponent("test_all_of.json")
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: How about making these lines into a private helper function?
```
private func loadFixture(named fileName: String) throws -> String {
    let url = testFixtureFolder.appendingPathComponent(fileName)
    return try String(contentsOfFile: url.Path, encoding: .utf8)
}
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 9]
```diff
@@ -1,20 +1,129 @@
...
+        let jsonString = try! NSString(contentsOfFile: url.path, encoding: String.Encoding.utf8.rawValue) as String
+        let swagger = try! Swagger(JSONString: jsonString)
+        
+        guard
+            let baseDefinition = swagger.definitions.first(where: {$0.name == "TestAllOfBase"}),
+            case .object(let baseSchema) = baseDefinition.structure else {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: When you have a multi-line guard statement, break the trailing curly brace onto the next line so Xcode doesn't double-indent the next line.

Comment (AttilaTheFun) [reply Tests/SwaggerParserTests/SwaggerParserTests.swift]: If you want to get fancy...
```
guard case .some(.object(let baseSchema)) = swagger.definitions.first(where: {$0.name == "TestAllOfBase"}) else {
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 10]
```diff
@@ -1,20 +1,129 @@
...
+                XCTFail("TestAllOfBase is not an object schema."); return
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: De-indent this

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 11]
```diff
@@ -1,20 +1,129 @@
...
         }
+        
+        SwaggerParserTests.validate(testAllOfBaseSchema: baseSchema)
+        swagger.definitions.validate(testAllOfChild: "TestAllOfFoo", withPropertyNames: ["foo"])
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: I think all three of these could be separate tests.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 12]
```diff
@@ -1,20 +1,129 @@
...
+        
+        if case .allOf(let schema) = definition.structure {
+            XCTAssertEqual(schema.schemas.count, 2)
+            
+            schema.schemas.forEach {
+                switch $0 {
+                case .object(let childSchema):
+                    child = childSchema
+                case .structure(let structure):
+                    guard case .object(let baseSchema) = structure.structure else {fatalError()}
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Would prefer we don't crash in test cases. Instead can we throw the error up to the top level test case and wrap it in a try/catch? Then we can catch + print the error and fail the test.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 13]
```diff
@@ -1,20 +1,129 @@
...
-        ("testUber", testUber),
-    ]
+fileprivate extension SwaggerParserTests {
+    /// Gets the base schema and child schema from a definition that defines an
+    /// `allOf` with one $ref (the base class) and one object schema.
+    class func getBaseAndChildSchemas(withDefinition definition: Structure<Schema>) -> (base: ObjectSchema?, child: ObjectSchema?) {
+        var base: ObjectSchema?
+        var child: ObjectSchema?
+        
+        if case .allOf(let schema) = definition.structure {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Let's call this allOfSchema

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 14]
```diff
@@ -1,20 +1,129 @@
...
+                case .object(let childSchema):
+                    child = childSchema
+                case .structure(let structure):
+                    guard case .object(let baseSchema) = structure.structure else {fatalError()}
+                    base = baseSchema
+                default: fatalError()
+                }
+            }
+        }
+        return (base: base, child: child)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: How about we make base and child implicitly unwrapped optionals and then throw an error if either is nil before we return?

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 15]
```diff
@@ -1,20 +1,129 @@
...
+        
+        properties.forEach { property in
+            XCTAssertNotNil(schema.properties.first(where: {$0.key == property}))
+        }
+    }
+    
+    class func validate(thatParameter parameter: Parameter, named parameterName: String, isAnObjectNamed objectName: String, withPropertyName objectPropertyName: String) {
+        guard case .body(_, let schema) = parameter else {
+            XCTFail("\(parameterName) is not a .body."); return
+        }
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: I like how you structured this with the series of guard statements.
A couple of nits:
- Newline after each closing curly preceding a statement.
- Let's avoid combining statements into one line like this - semicolons are the devil!
- XCTFail is a void function. You can just do:
```
return XCTFail(...)
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 16]
```diff
@@ -1,20 +1,129 @@
...
+        
+        XCTAssertTrue(object.properties.contains(where: {$0.key == objectPropertyName}))
+    }
+    
+    class func validate(thatChildSchema childSchema: Schema, named childName: String, withProperties childProperties: [String], hasParentNamed parentName: String, withProperties parentProperties: [String]) {
+        guard case .allOf(let childAllOf) = childSchema else {
+            XCTFail("\(childName) is not an allOf."); return
+        }
+        XCTAssertEqual(childAllOf.schemas.count, 2)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: \n

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 17]
```diff
@@ -1,20 +1,129 @@
...
+        
+        guard
+            let childsParent = childAllOf.schemas.first,
+            case .structure(let childsParentStructure) = childsParent,
+            childsParentStructure.name == parentName,
+            case .object(let childsParentSchema) = childsParentStructure.structure else {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Opening curly on newline, de-indent

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 18]
```diff
@@ -1,20 +1,129 @@
...
+                XCTFail("\(childName)'s parent is not a Structure<Schema.object>"); return
+        }
+        SwaggerParserTests.validate(thatSchema: childsParentSchema, named: parentName, hasRequiredProperties: parentProperties)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: \n

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 19]
```diff
@@ -1,20 +1,129 @@
...
+            }
+        }
+        return (base: base, child: child)
+    }
+}
+
+/// MARK: Validation functions
+
+fileprivate extension SwaggerParserTests {
+    class func validate(testAllOfBaseSchema schema: ObjectSchema) {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: I don't think these functions need to be class functions. I'd honestly just make them private functions (defined at file scope).

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 20]
```diff
@@ -1,20 +1,129 @@
...
+                XCTFail("\(childName)'s parent is not a Structure<Schema.object>"); return
+        }
+        SwaggerParserTests.validate(thatSchema: childsParentSchema, named: parentName, hasRequiredProperties: parentProperties)
+        
+        guard
+            let child = childAllOf.schemas.last,
+            case .object(let childSchema) = child else {
+                XCTFail("child is not a Structure<Schema.object>"); return
+        }
+        SwaggerParserTests.validate(thatSchema: childSchema, named: childName, hasRequiredProperties: childProperties)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: \n

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 21]
```diff
@@ -1,20 +1,129 @@
...
+            case .structure(let childsParentStructure) = childsParent,
+            childsParentStructure.name == parentName,
+            case .object(let childsParentSchema) = childsParentStructure.structure else {
+                XCTFail("\(childName)'s parent is not a Structure<Schema.object>"); return
+        }
+        SwaggerParserTests.validate(thatSchema: childsParentSchema, named: parentName, hasRequiredProperties: parentProperties)
+        
+        guard
+            let child = childAllOf.schemas.last,
+            case .object(let childSchema) = child else {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: curly on newline, de-indent

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 22]
```diff
@@ -1,20 +1,129 @@
...
+        }
+        SwaggerParserTests.validate(thatSchema: childSchema, named: childName, hasRequiredProperties: childProperties)
+    }
+}
+
+/// MARK: Swagger Definitions Extension
+
+fileprivate extension Array where Element == Structure<Schema> {
+    func validate(testAllOfChild name: String, withPropertyNames propertyNames: [String]) {
+        let testAllOfChild = self.first(where: {$0.name == name})!
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Let's not do force unwraps - I like what you were doing earlier with the guard statements where the else has XCTFail

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 23]
```diff
@@ -1,20 +1,129 @@
...
+    }
+    
+    class func validate(thatParameter parameter: Parameter, named parameterName: String, isAnObjectNamed objectName: String, withPropertyName objectPropertyName: String) {
+        guard case .body(_, let schema) = parameter else {
+            XCTFail("\(parameterName) is not a .body."); return
+        }
+        guard case .structure(let structure) = schema else {
+            XCTFail("\(parameterName)'s schema is not a .structure."); return
+        }
+        XCTAssertEqual(structure.name, objectName)
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: \n

File: Tests/SwaggerParserTests/SwaggerParserTests.swift (hunk 14 above)
Comment (AttilaTheFun) [reply Tests/SwaggerParserTests/SwaggerParserTests.swift]: Also newline before statement

File: Sources/AllOfSchema.swift [hunk 24]
```diff
@@ -0,0 +1,35 @@
+import ObjectMapper
+
+public struct AllOfSchema {
+    public let metadata: Metadata
+    
+    public let schemas: [Schema]
```
Comment (AttilaTheFun) [inline Sources/AllOfSchema.swift]: I think we should call this subschemas or something less generic

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 25]
```diff
@@ -1,20 +1,129 @@
...
+    /// `allOf` with one $ref (the base class) and one object schema.
+    class func getBaseAndChildSchemas(withDefinition definition: Structure<Schema>) -> (base: ObjectSchema?, child: ObjectSchema?) {
+        var base: ObjectSchema?
+        var child: ObjectSchema?
+        
+        if case .allOf(let schema) = definition.structure {
+            XCTAssertEqual(schema.schemas.count, 2)
+            
+            schema.schemas.forEach {
+                switch $0 {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Please don't use $0 if your closure is more than one line.
```
allOfSchema.subschemas.forEach { schema in
    switch schema {
    ...
    }
}
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 26]
```diff
@@ -1,20 +1,129 @@
...
+                case .object(let childSchema):
+                    child = childSchema
+                case .structure(let structure):
+                    guard case .object(let baseSchema) = structure.structure else {fatalError()}
+                    base = baseSchema
+                default: fatalError()
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: I dislike putting statements on the same line as the colon. Also let's not fatalError() - how about throwing an error or doing XCTFail

**Review by AttilaTheFun — CHANGES_REQUESTED** [review]
Looks pretty good overall!

**Comment by AttilaTheFun** [comment]
@n8chur thanks! I just finished going through your PR and leaving comments. It looked pretty good overall, mostly just stylistic / nitpicky things. FYI the reason I don't like crashing in tests is that once you have a decent suite of tests, if one of the early ones crashes it doesn't end up running your full test suite.

**Comment by yonaskolb** [comment]
Just thought I'd chime in here. You can also mark a test as throws and XCTest will automatically fail if something in the test throws. Saves you from try catching, and means you can easily mark any helper functions as throws as well

**Comment by AttilaTheFun** [comment]
@yonaskolb oh cool! Didn't know that. @n8chur we should definitely take advantage of that,

File: Tests/SwaggerParserTests/SwaggerParserTests.swift (hunk 9 above)
Comment (n8chur) [reply Tests/SwaggerParserTests/SwaggerParserTests.swift]: That is fancy! I didn't realize you could do that. That being said I do think keeping them new line separated helps with readability.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift (hunk 11 above)
Comment (n8chur) [reply Tests/SwaggerParserTests/SwaggerParserTests.swift]: This test was intended to validate that the swagger.json file is being parsed correctly which implies that each object should be validated within this test. Should I break it out into a separate test per schema being parsed? 

If we do break them out, we may want to consider making separate test files for each feature. 

I feel like a BDD framework would make breaking things out more granularly easier to manage (with `context`s and `describe` scopes).

**Comment by n8chur** [comment]
@yonaskolb That's awesome! Thanks for the tip!

File: Sources/AllOfSchema.swift [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+        metadata = try MetadataBuilder(map: map)
+        
+        schemaBuilders = try map.value("allOf")
+    }
+    
+    func build(_ swagger: SwaggerBuilder) throws -> AllOfSchema {
+        let subschemas = try schemaBuilders.map { try $0.build(swagger) }
+        return AllOfSchema(metadata: try self.metadata.build(swagger), subschemas: subschemas)
+    }
+
```
Comment (AttilaTheFun) [inline Sources/AllOfSchema.swift]: I prefer not to have trailing whitespace - can you remove this newline?

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 28]
```diff
@@ -1,20 +1,159 @@
...
+        try allOfSchema.subschemas.forEach { subschema in
+            switch subschema {
+            case .object(let childSchema):
+                child = childSchema
+                
+            case .structure(let structure):
+                guard case .object(let baseSchema) = structure.structure else {
+                    throw GetBaseAndChildSchemasError.badSubschemaType(subschema)
+                }
+                base = baseSchema
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Can you add a newline after the closing curly brace before the next statement? If the curly brace ends the last statement in that scope the newline isn't necessary though. Also I don't think the newlines separating the case statements are necessary. They're already indented.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 29]
```diff
@@ -1,20 +1,159 @@
...
+/// MARK: Helper Functions
+
+fileprivate extension SwaggerParserTests {
+    func fixture(named fileName: String) throws -> String {
+        let url = testFixtureFolder.appendingPathComponent(fileName)
+        return try String.init(contentsOf: url, encoding: .utf8)
+    }
+}
+    
+fileprivate enum GetBaseAndChildSchemasError: Error {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift:44]: let's add
```
case notAllOf
case incorrectSubschemaCount
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 30]
```diff
@@ -1,20 +1,159 @@
...
+                }
+                base = baseSchema
+                
+            default:
+                throw GetBaseAndChildSchemasError.badSubschemaType(subschema)
+            }
         }
     }
+    
+    guard base != nil else {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Don't use guard statements just to perform boolean checks - they should be unwrapping something usually.

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 31]
```diff
@@ -1,20 +1,159 @@
...
+    case badSubschemaType(Schema)
+}
+
+/// Gets the base schema and child schema from a definition that defines an
+/// `allOf` with one $ref (the base class) and one object schema.
+fileprivate func getBaseAndChildSchemas(withDefinition definition: Structure<Schema>) throws -> (base: ObjectSchema, child: ObjectSchema) {
+    var base: ObjectSchema!
+    var child: ObjectSchema!
+    
+    if case .allOf(let allOfSchema) = definition.structure {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: How about turning this into a guard statement that throws an error. Then you don't have to indent the rest of this if statement. I.e.:

```
guard case .allOf(let allOfSchema) = definition.structure else {
    throw GetBaseAndChildSchemasError.notAllOf
}

XCTAssertEqual(allOfSchema.subschemas.count, 2)

var base: ObjectSchema?
var child: ObjectSchema?

try allOfSchema.subschemas.forEach { subschema in
    switch subschema {
    case .object(let childSchema):
        child = childSchema
    case .structure(let structure):
        if case .object(let baseSchema) = structure.structure {
            base = baseSchema
        } else {
            throw GetBaseAndChildSchemasError.badSubschemaType(subschema)
        }
    default:
        throw GetBaseAndChildSchemasError.badSubschemaType(subschema)
    }
}

if let base = base, let child = child {
    return (base: base, child: child)
}

if base == nil {
    throw GetBaseAndChildSchemasError.missingBase
} else {
    throw GetBaseAndChildSchemasError.missingChild
}

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 32]
```diff
@@ -1,20 +1,159 @@
...
+fileprivate func validate(testAllOfBaseSchema schema: ObjectSchema) {
+    validate(thatSchema: schema, named: "TestAllOfBase", hasRequiredProperties: ["base", "test_type"])
+}
+
+fileprivate func validate(thatSchema schema: ObjectSchema, named name: String, hasRequiredProperties properties: [String]) {
+    XCTAssertEqual(schema.properties.count, properties.count)
+    XCTAssertEqual(schema.required, properties)
+    
+    properties.forEach { property in
+        XCTAssertNotNil(schema.properties.first(where: {$0.key == property}))
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: This is O(n^2) - even if it is for a test. How about:
```
let keys = Set(schema.properties.keys)
properties.forEach { XCTAssert(keys.contains($0) }
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 33]
```diff
@@ -1,20 +1,159 @@
...
+    }
+}
+
+fileprivate func validate(thatParameter parameter: Parameter, named parameterName: String, isAnObjectNamed objectName: String, withPropertyName objectPropertyName: String) {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Same deal - let's not repeat the type information in the signature.
```
fileprivate func validate(that parameter: Parameter, ...
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 34]
```diff
@@ -1,20 +1,159 @@
...
+        return XCTFail("\(parameterName)'s schema is not a .structure.")
+    }
+    
+    XCTAssertEqual(structure.name, objectName)
+    
+    guard case .object(let object) = structure.structure else {
+        return XCTFail("\(parameterName)'s schema's structure is not an .object.")
+    }
+    
+    XCTAssertTrue(object.properties.contains(where: {$0.key == objectPropertyName}))
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Do we need the where: here? Can't we just use the trailing closure syntax? In either case, let's put spaces inside the curly braces:
```
XCTAssertTrue(object.properties.contains { $0.key == objectPropertyName })

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 35]
```diff
@@ -1,20 +1,159 @@
...
+}
+
+fileprivate func validate(thatChildSchema childSchema: Schema, named childName: String, withProperties childProperties: [String], hasParentNamed parentName: String, withProperties parentProperties: [String]) {
+    guard case .allOf(let childAllOf) = childSchema else {
+        XCTFail("\(childName) is not an allOf."); return
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Please don't use the semicolon to connect statements - XCTFail is a void function anyways so you can:
```
return XCTFail("\(childName) is not an allOf.")
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 36]
```diff
@@ -1,20 +1,159 @@
...
+        case .object(let childsParentSchema) = childsParentStructure.structure else
+    {
+        return XCTFail("\(childName)'s parent is not a Structure<Schema.object>")
+    }
+    
+    validate(thatSchema: childsParentSchema, named: parentName, hasRequiredProperties: parentProperties)
+    
+    guard
+        let child = childAllOf.subschemas.last,
+        case .object(let childSchema) = child else
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Can't you fit this all onto one line? Usually I set my guide at 110 characters (that's about the maximum Github can display cleanly in the diff view).
```
guard let child = childAllOf.subschemas.last, case .object(let childSchema) = child else {
```

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 37]
```diff
@@ -1,20 +1,159 @@
...
+    
+    validate(thatSchema: childSchema, named: childName, hasRequiredProperties: childProperties)
+}
 
+/// MARK: Swagger Definitions Extension
 
-    static var allTests = [
-        ("testUber", testUber),
-    ]
+fileprivate func validate(thatDefinitions definitions: [Structure<Schema>], containsTestAllOfChild name: String, withPropertyNames propertyNames: [String]) throws {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Let's not repeat type in the signature:
```
fileprivate func validate(that definitions ...

File: Tests/SwaggerParserTests/SwaggerParserTests.swift [hunk 38]
```diff
@@ -1,20 +1,159 @@
...
+    guard let testAllOfChild = definitions.first(where: {$0.name == name}) else {
```
Comment (AttilaTheFun) [inline Tests/SwaggerParserTests/SwaggerParserTests.swift]: Spaces around the curly braces:
```
{ $0.name == name }

**Review by AttilaTheFun — CHANGES_REQUESTED** [review]
Looking good! I just added a few more comments, mostly style related. I would refactor the getBaseAndChildSchemas function like I proposed, however.

**Review by AttilaTheFun — APPROVED** [review]
Awesome! Thanks so much!

**Comment by n8chur** [comment]
Thanks for viewing that and getting it merged!

**Comment by AttilaTheFun** [comment]
No problem! Feel free to create another PR for the other stuff you mentioned

**Pull request merged**